custom/icds_reports/reports/service_delivery_dashboard_data.py

```diff
@@ -0,0 +1,352 @@
+        'data': [],
+    }
+
+    def update_all_row(first_dict, second_dict):
+        for key, value in first_dict.items():
+            # excluding location and percentage fields
+            if key in values:
+                first_dict[key] = format_data_not_entered_to_zero(first_dict[key]) +\
+                                  format_data_not_entered_to_zero(second_dict[key])
+        return first_dict
+
+    def get_value_or_data_not_entered(source, field):
+        value = source.get(field)
+        if value is None:
+            return DATA_NOT_ENTERED
+        return value
+
+    def get_thr_percents(base_dict, row_data):
+        base_dict['thr_0_days_val'] = percent_or_not_entered(row_data['thr_0_days'],
+                                                             row_data['thr_eligible'])
+        base_dict['thr_1_7_days_val'] = percent_or_not_entered(row_data['thr_1_7_days'],
+                                                               row_data['thr_eligible'])
+        base_dict['thr_8_14_days_val'] = percent_or_not_entered(row_data['thr_8_14_days'],
+                                                                row_data['thr_eligible'])
+        base_dict['thr_15_20_days_val'] = percent_or_not_entered(row_data['thr_15_20_days'],
+                                                                 row_data['thr_eligible'])
+        base_dict['thr_21_24_days_val'] = percent_or_not_entered(row_data['thr_21_24_days'],
+                                                                 row_data['thr_eligible'])
+        base_dict['thr_25_days_val'] = percent_or_not_entered(row_data['thr_25_days'],
+                                                              row_data['thr_eligible'])
+        return base_dict
+
+    def get_sn_percents(base_dict, row_data):
+        base_dict['lunch_0_days_val'] = percent_or_not_entered(row_data['lunch_0_days'],
+                                                               row_data['pse_eligible'])
+        base_dict['lunch_1_7_days_val'] = percent_or_not_entered(row_data['lunch_1_7_days'],
+                                                                 row_data['pse_eligible'])
+        base_dict['lunch_8_14_days_val'] = percent_or_not_entered(row_data['lunch_8_14_days'],
+                                                                  row_data['pse_eligible'])
+        base_dict['lunch_15_20_days_val'] = percent_or_not_entered(row_data['lunch_15_20_days'],
+                                                                   row_data['pse_eligible'])
+        base_dict['lunch_21_24_days_val'] = percent_or_not_entered(row_data['lunch_21_24_days'],
+                                                                   row_data['pse_eligible'])
+        base_dict['lunch_25_days_val'] = percent_or_not_entered(row_data['lunch_25_days'],
+                                                                row_data['pse_eligible'])
+        return base_dict
+
+    def get_pse_percents(base_dict, row_data):
```

Comment:
could you take these nested functions out and make them private to this file like
`_get_pre_percents()` then call that here that way length of this function would reduce

Reply:
Moreover could these functions be made more generalized since all the fields are similar looking. So , There could be only one function such as 

```
_get_service_percents(base_dict, row_data, service_name):
    base_dict[f'{service_name}_0_days_val'] = percent_or_not_entered(row_data[f'{service_name}_0_days'],
                                                             row_data['pse_eligible'])
    base_dict[f'{service_name}_1_7_days_val'] = percent_or_not_entered(row_data[f'{service_name}_1_7_days'],
                                                               row_data[f'{service_name}_eligible'])

_get_service_percents(x,y, 'thr')
_get_service_percents(x,y,'pse')
```

using above function would reduce these 3 function to just one function

Reply:
thought of it but did not implement as it makes the fields coupled. Changed it now